ppa/archive/management/commands/generate_textcorpus.py

```diff
@@ -2,199 +2,238 @@
-        for doc_id in self.doc_ids:
-            # get saved filename and found metadata for this document
-            fn,meta = self._save_doc(doc_id)
+    #### SOLR ####
 
-            # if we saved, update progress bar desc
-            if fn: pbar.set_description(f'{pdesc}: {fn}')
+    @cached_property
+    def query_set(self):
+        """
+        A cached representation of the query set;
+        it discourages opening multiple connections to database.
+        """
+        return SolrQuerySet()
 
-            # tick
-            pbar.update()
+    def iter_solr(self, item_type="page"):
+        """
+        Iterate over solr documents of a certain `item_type`.
+        """
 
-            # add this doc's meta to metadata
-            metadata.append(meta)
-        pbar.close()
+        # get params
+        batch_size, lim = self.batch_size, self.doclimit
 
-        # save metadata csv
-        with open(self.path_metadata,'w') as of:
-            json.dump(metadata, of, indent=2)
-        print(f'Saved metadata to: {self.path_metadata}')
+        # search for this item type, order by id
+        qset = self.query_set.search(item_type=item_type).order_by("id")
 
+        # if we're looking for pages, rename those fields
+        if item_type == "page":
+            qset = qset.only(*self.PAGE_FIELDLIST)
+        elif item_type == "work":
+            qset = qset.only(*self.WORK_FIELDLIST)
 
+        # get the total count for this query
+        total = qset.count()
 
+        # if we want fewer than that, decrease "total"
+        # (this has the effect of pulling from solr only what we need,
+        # since we limit how many rows we pull by this amount)
+        if lim and total > lim:
+            total = lim
 
+        # if total smaller than batch size, decrease batch size
+        if batch_size > total:
+            batch_size = total
 
-@contextmanager
-def logging_disabled(highest_level=logging.CRITICAL):
-    """Quick way to suppress solr logs as we iterate. Taken from https://gist.github.com/simon-weber/7853144"""
-    previous_level = logging.root.manager.disable
-    logging.disable(highest_level)
-    try: yield
-    finally: logging.disable(previous_level)
+        # define a generator to iterate solr with
+        iterr = (
+            result
+            for step in range(0, total, batch_size)
+            for result in qset[step : step + batch_size]
+        )
 
+        # if progress bar wanted, tack one on
+        if self.progress:
+            iterr = progressbar(iterr, max_value=total)
 
+        # yield from this generator, progress bar or no
+        yield from iterr
 
+    def iter_works(self):
+        """
+        Simply calls `.iter_solr()` with `item_type` as `'work'`
+        """
+        for d in self.iter_solr(item_type="work"):
+            # source does not need to be a list, will only be one
+            assert len(d["source"]) == 1
+            d["source"] = d["source"][0]
```

Comment:
I had been thinking of using the aliased solr queryset (which lets you rename fields)  in parasolr, which now has a method for updating the result document before it returns. I kind of suspect your generator solution is more efficient, though.